Ship Cody Improvements for 5.3 release (PLG & Enterprise) (#69)
sourcegraph/docs: docs/cody/clients/enable-cody-enterprise.mdx

# Enabling Cody on Sourcegraph Enterprise

<p className="subtitle">This guide will walk you through the steps to install and set up Cody with your Sourcegraph Enterprise instance.</p>

Cody enhances your coding experience by providing intelligent code suggestions, context-aware completions, and advanced code analysis.

<LinkCards>
   <LinkCard href="https://about.sourcegraph.com/cody/pricing" imgSrc="https://sourcegraph.com/.assets/img/sourcegraph-mark.svg" imgAlt="Cody Enterprise" title="Cody Enterprise" description="Get in touch with our team to try Cody for Sourcegraph Enterprise." />
</LinkCards>

## Cody Enterprise features

To cater to your Enterprise requirements, Cody Enterprise offers the following features:

### IDE Token Expiry

> <Callout type= "info">IDE Token Expiry is supported with VS Code and JetBrains extensions.</Callout>

Reviewer: This applies to access tokens in general not just when in use with Cody. I previously updated the site config docs to show these options.

Author: Thanks I'll remove the callout.

### Guardrails

<Callout type= "note">Guardrails for public code is currently in Beta stage and is supported with VS Code and JetBrains IDEs extensions.</Callout>

Open source attribution guardrails for public code reduce the exposure to copyrighted code, commonly called copyright guardrails. This involves the implementation of a verification mechanism within Cody to ensure that any code generated by the platform does not replicate open source code.

Guardrails for public code are available to all Sourcegraph Enterprise instances and are **disabled** by default. You can enable it from the **Site configuration** or the IDE extension settings. It only matches code snippets that are at least **10 lines** or longer, and it spans **290,000** search results.

### Admin Controls

<Callout type="note">Admin Controls is supported with VS Code IDE extension.</Callout>

Sourcegraph account admins have selective control over users' access to Cody Enterprise, which is now managed via the Sourcegraph role-based access control system. This provides a more intuitive user interface for assigning permission to use Cody.

> Using the access controls, you can also configure which Cody capabilities users can have access to (for example, chat, completion, or commands).

Reviewer: ```suggestion

```
This only got implemented for VS Code so I don't think we should call it a capability

### Analytics

<Callout type= "note">Admin Controls is supported with VS Code IDE extension and on the latest versions of JetBrains IDE.</Callout>

> Cody Enterprise users get a clear view of usage analytics for their instance on a self-service basis. A separately managed cloud service for Cody analytics (for example, `cody-analytics.sourcegraph.com`) handles user auth, gets metrics data from Sourcegraph's BigQuery instance, and visualizes the metrics data.

Reviewer 2: That's the only prod instance of the Cody Analytics service. I think we can remove the "for example" part.

| **Metric Type** |                                                                                                                                                                         **What is measured?**                                                                                                                                                                          |
| --------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Active users    | - Total active users <br /> - Average daily users <br /> - Average no. of days each user used Cody (of last 30 days) <br /> - Cody users by day (last 30 days) <br /> - Cody users by month (last two months) <br /> - Cody users by number of days used                                                                                                               |

> | Completions     | - Total accepted completions <br /> - Minutes saved per completion (hardcoded on our end) <br /> - Hours saved by completions (computed based on the previous two) <br /> - Cody completions by day <br /> - Completions acceptance rate <br /> - Weighted completions acceptance rate <br /> - Average completion latency <br /> - Acceptance rate by language <br /> |

Reviewer 2: I think we can remove "(hardcoded on our end)" here and after. It was mentioned for additional context on the internal meetings.

> | Completions     | - Total accepted completions <br /> - Minutes saved per completion (hardcoded on our end) <br /> - Hours saved by completions (computed based on the previous two) <br /> - Cody completions by day <br /> - Completions acceptance rate <br /> - Weighted completions acceptance rate <br /> - Average completion latency <br /> - Acceptance rate by language <br /> |

Reviewer 2: ```suggestion
| Completions     | - Total accepted completions <br /> - Minutes saved per completion <br /> - Hours saved by completions <br /> - Cody completions by day <br /> - Completions acceptance rate <br /> - Weighted completions acceptance rate <br /> - Average completion latency <br /> - Acceptance rate by language <br /> |
```

Reviewer 2: The removed parts were mentioned for additional context on the internal meetings.

> | Chat            | - Total chat events <br /> - Minutes saved per chat <br /> - Hours saved by chats (computed based on the previous two) <br /> - Cody chats by day                                                                                                                                                                                                                      |

Reviewer 2: ```suggestion
| Chat            | - Total chat events <br /> - Minutes saved per chat <br /> - Hours saved by chats <br /> - Cody chats by day                                                                                                                                                                                                                      |
```

> | Commands        | - Total command events <br /> - Minutes saved per command <br /> - Hours saved by commands (computed based on the previous two) <br /> - Cody commands by day <br /> - Most used commands                                                                                                                                                                              |

Reviewer 2: ```suggestion
| Commands        | - Total command events <br /> - Minutes saved per command <br /> - Hours saved by commands <br /> - Cody commands by day <br /> - Most used commands                                                                                                                                                                              |
```

To enable Cody Analytics:

- Create an account on [Sourcegraph Accounts](https://accounts.sourcegraph.com/)
- A user already having an account on Sourcegraph.com gets automatically migrated to Sourcegraph Accounts. Users can sign in to Cody Analytics using their email and password
- Users without a Sourcegraph.com account, please get in touch with support to request the creation of a new account. Once the account is set up, users will be able to sign in using their email and password

> - Map your user account to a Sourcegraph instance, and this gives you access to Cody's analytics

Author: The readme link gives me a 404. @taras-yemets

> - Map your user account to a Sourcegraph instance, and this gives you access to Cody's analytics

Reviewer 2: @MaedahBatool, I just updated repo permissions. You should be able to access it now.

To enable StarCoder go to **Site admin > Site configuration** (`/site-admin/configuration`) and change the `completionModel`:

```json
{
  // [...]
  "cody.enabled": true,
  "completions": {
    "provider": "sourcegraph",
    "completionModel": "fireworks/starcoder"
  }
}
```

Cody Extension will automatically pick up this change when they connect to your Enterprise instance.

It's recommended that every instance admin not using a third-party LLM provider makes this change and we are planning to make this the default in a future release.

### Multi-repo Context

<Callout type= "info">Multi-repo context for Sourcegraph Enterprise is supported with VS Code and JetBrains IDEs extensions.</Callout>

Cody Enterprise allows up to 10 repos to support multi-repo context for users on IDE extensions. These can be added from the enhanced context selector window. Multi-repo context can search and retrieve context from multiple repositories responding to user questions, opening up more complex use cases. Enterprise users can ask questions requiring broader knowledge from across a codebase.

> ### Cody ignore

Reviewer: Cody Ignore support is experimental.

> Enterprise instances can flag repositories and files that they don't want to be used by Cody as context for autocomplete and chat, safeguarding sensitive information from being returned to users.

Reviewer: We do not allow marking repositories as unavailable for Cody.  The experimental support is only to exclude some/all files within a repo based on the ignore file at `.cody/ignore`

</LinkCards>

## Cody Enterprise features

To cater to your Enterprise requirements, Cody Enterprise offers the following features:

### IDE Token Expiry

Site administrators can customize the time duration of the access token used by other users on the Sourcegraph Enterprise instance to connect Cody from their IDEs via the **Site admin** page. Administrators can choose from various options, including 7, 14, 30, 60, and 90 days.

![ide-token-expiry](https://storage.googleapis.com/sourcegraph-assets/Docs/token-expiry.png)

### Guardrails

<Callout type= "note">Guardrails for public code is currently in Beta stage and is supported with VS Code and JetBrains IDEs extensions.</Callout>

Open source attribution guardrails for public code reduce the exposure to copyrighted code, commonly called copyright guardrails. This involves the implementation of a verification mechanism within Cody to ensure that any code generated by the platform does not replicate open source code.

> Guardrails for public code are available to all Sourcegraph Enterprise instances and are **disabled** by default. You can enable it from the **Site configuration** or the IDE extension settings. It only matches code snippets that are at least **10 lines** or longer, and it spans **290,000** search results.

Reviewer 3: ```suggestion
Guardrails for public code are available to all Sourcegraph Enterprise instances and are **disabled** by default. You can enable it from the **Site configuration**. It only matches code snippets that are at least **10 lines** or longer, and the search corpus **290,000** open source repositories.
```

> Guardrails for public code are available to all Sourcegraph Enterprise instances and are **disabled** by default. You can enable it from the **Site configuration** or the IDE extension settings. It only matches code snippets that are at least **10 lines** or longer, and it spans **290,000** search results.

Reviewer 3: We can also add that guardrails is enabled by setting `"attribution.enabled": true` in site config.

### Admin Controls

> <Callout type="note">Admin Controls is supported with VS Code IDE extension.</Callout>

Reviewer 3: ```suggestion
<Callout type="note">Admin Controls is supported with VS Code and JetBrains IDE extension.</Callout>
```

right? @chwarwick @slimsag

Reviewer: Yes, it's not actually client specific  just a True/False if the user has access

Sourcegraph account admins have selective control over users' access to Cody Enterprise, which is now managed via the Sourcegraph role-based access control system. This provides a more intuitive user interface for assigning permission to use Cody.

### Analytics

> <Callout type= "note">Admin Controls is supported with VS Code IDE extension and on the latest versions of JetBrains IDE.</Callout>

Reviewer 3: ```suggestion
```

I think this call out mentions Admin controls.

Author: Oh yes my bad, I'll fix it.

> <Callout type= "note">Admin Controls is supported with VS Code IDE extension and on the latest versions of JetBrains IDE.</Callout>

Reviewer 4: Is this supposed to say "Analytics is supported ...."

Cody Enterprise allows up to 10 repos to support multi-repo context for users on IDE extensions. These can be added from the enhanced context selector window. Multi-repo context can search and retrieve context from multiple repositories responding to user questions, opening up more complex use cases. Enterprise users can ask questions requiring broader knowledge from across a codebase.

### Enhanced context selector

<Callout>Enhanced context selector is currently supported on VS Code and JetBrains IDE extensions.</Callout>

The **Enhanced context selector** panel allows Cody Enterprise users to add up to **10 repos** to search across multiple repositories. This will enable users to ask questions requiring broader knowledge from across a codebase.

Create a new chat in VS Code and click the **Enhanced Context** icon in the chat type field. Click the **Choose Repositories...** button and select up to 10 repositories from the list.

![enhanced-context-enterprise](https://storage.googleapis.com/sourcegraph-assets/Docs/enhanced-context-enterprise.png)

> ### Cody ignore
> 
> <Callout type="info">Cody Ignore is currently in the Experimental stage and is only supported in VS Code extension.</Callout>
> 
> Users can choose files and folders to exclude as context for Cody. It works as follows:
> 
> - `.cody/ignore` allows users to use the familiar `gitignore` file pattern to define files that they do not want Cody to use as context
> - The purpose of `.cody/ignore` is to ensure that certain files are not shared with third parties LLM provider
> - Executing a command on an ignored file will trigger a system pop-up to notify you that Cody is ignoring the file

Reviewer: ```suggestion
```
Cody ignore is still behind the unstable flag in VS Code it feels too soon to document how it works, because following these instructions will not activate it.
